FirebaseDeviceStore/FirebaseDeviceStore.swift

```diff
@@ -120,78 +131,39 @@ public class FirebaseDeviceStore: NSObject, MessagingDelegate {
         subscribed = false;
     }
 
-    private func deleteDevice(_ userId: String) {
+    private func deleteDevice(_ userId: String, completion: @escaping (Error?) -> Void) {
         let docRef = userRef(userId);
 
-        firestore.runTransaction({ (transaction, errorPointer) -> Any? in
-            let doc: DocumentSnapshot;
-            do {
-                try doc = transaction.getDocument(docRef);
-            } catch let fetchError as NSError {
-                errorPointer?.pointee = fetchError
-                return nil;
-            }
-
-            if (doc.exists) {
-                var devices = self.getDevices(doc);
-                // Remove the old device
-                devices = self.removeCurrentDevice(devices);
-                // Update the document
-                transaction.updateData([self.DEVICES_FIELD: devices], forDocument: docRef);
-            } else {
-                let devices = self.createUserDevices(userId, nil);
-                transaction.setData(devices, forDocument: docRef);
+        docRef.getDocument { (doc, error) in
+            if (doc != nil) {
+                if (doc!.exists) {
+                    return docRef.updateData([FieldPath.init([self.DEVICES_FIELD, self.getDeviceId()]): FieldValue.delete()], completion: completion);
+                }
+                return completion(nil);
             }
-            return nil;
-        }) { (object, error) in
-            // TODO: Logging
+            return completion(error);
         }
     }
 
-    private func updateDevice(_ userId: String, _ token: String) {
+    private func updateDevice(_ userId: String, _ token: String, completion: @escaping (Error?) -> Void) {
         let docRef = userRef(userId);
 
-        firestore.runTransaction({ (transaction, errorPointer) -> Any? in
-            let doc: DocumentSnapshot;
-            do {
-                try doc = transaction.getDocument(docRef);
-            } catch let fetchError as NSError {
-                errorPointer?.pointee = fetchError
-                return nil;
-            }
-            if (doc.exists) {
-                var devices = self.getDevices(doc);
-                if (self.containsCurrentDevice(devices)) {
-                    // Update the device token if it already exists
-                    self.updateCurrentDevice(devices, token);
+        docRef.getDocument { (doc, error) in
+            if (doc != nil) {
+                if (doc!.exists) {
+                    let deviceId = self.getDeviceId();
+                    return docRef.updateData([FieldPath.init([self.DEVICES_FIELD, self.getDeviceId()]): self.createDevice(deviceId, token)], completion: completion);
                 } else {
-                    devices.append(self.createCurrentDevice(token));
+                    return docRef.setData(self.createUserDevices(userId, token), completion: completion);
```

Comment:
The way you've written this (now that there's no transaction), there's technically a race condition since another instance of the app could theoretically create docRef *after* you call getDocument() but *before* you call setData() and so your setData() would clobber that data.

To be more correct (and less code), you could skip calling getDocument() and just always use `setData(..., merge: true)` to merge the new values into the existing document (and create it if it doesn't already exist):

```
docRef.setData([USER_ID_FIELD: userId, DEVICES_FIELD: [self.getDeviceId(): self.createDevice(deviceId, token)], merge: true, completion: completion)
```

Also, setData() returns void so I don't think you need `return`?